Add support for parameters in `Target::instruction_supported` in Rust.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #14957

- Add the `parameters` and `check_angle_bounds` arguments to the Rust-native `instruction_supported` method for `Target`. This enables users to use parameter checks with rust native structure.
- The following passes have been updated to use this path:
    - `ConsolidateBlocks`.
    - `GateDirection`.
    - `HighLevelSynthesis`.
    - `GatesInBasis`.
- Add docstring to rust native `instruction_supported`.


### Details and comments

raynelfss committed:
Initial: Add support for parameters in `instruction_supported` in Rust.
- Add the `parameters` and `check_angle_bounds` arguments to the Rust-native `instruction_supported` method for `Target`. This enables users to use parameter checks with rust native structure.
- The following passes have been updated to use this path:
    - `ConsolidateBlocks`.
    - `GateDirection`.
    - `HighLevelSynthesis`.
    - `GatesInBasis`.
- Add docstring to rust native `instruction_supported`.

eliarbel commented on crates/transpiler/src/passes/gate_direction.rs:
```diff
@@ -184,29 +184,13 @@ pub fn fix_direction_coupling_map(
-            }
-        }
-        target.instruction_supported(inst.op.name(), &qargs)
+        target.instruction_supported(inst.op.name(), qargs, inst.params_view(), true)
```
Starting in https://github.com/Qiskit/qiskit/pull/14406, `py_instruction_supported` was called with passing `false` to the `check_angle_bounds` param. I believe your change is correct (i.e. passing `true`), but I wanted to clarify anyway that we are not missing anything. @mtreinish was there a reason to call it with angle bound check disabled?

mtreinish replied:
TBH, I am not 100% sure I remember my intent. I think based on context and @eliarbel's comment I didn't want to force a gate flip here if the directionality was correct but the angle invalid. I assume I was thinking that we want to save the angle wrapping for the pass handling that and only have this check that the gate direction/qargs are correct at this point.

eliarbel replied:
> I was thinking that we want to save the angle wrapping for the pass handling that and only have this check that the gate direction/qargs are correct at this point.

I think you're right and it's probably better to restrict the pass to only address the directionality aspect and defer angle normalization to `WrapAngles`.

@raynelfss ⬆️

raynelfss replied:
Reverted in 9094f79941127eaaf054d8471a26c0f0d86ec20b

eliarbel commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1396,43 +1365,72 @@ impl Target {
+                        let qarg_set: HashSet<&PhysicalQubit> = qargs_as_vec.iter().collect();
+                        return qargs_as_vec
+                            .iter()
+                            .all(|qarg| qarg.0 <= self.num_qubits.unwrap_or_default())
```
nitpicking: really negligible performance-wise, but `self.num_qubits.unwrap_or_default()` should be cached outside the closure.

eliarbel commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1396,43 +1365,72 @@ impl Target {
+                        return qargs_as_vec
+                            .iter()
+                            .all(|qarg| qarg.0 <= self.num_qubits.unwrap_or_default())
+                            && qarg_set.len() == qargs_as_vec.len();
```
Why do we need to check for duplicates?

raynelfss replied:
I only did it because we did this initially but I didn't give it much thought. I removed it and everything seems to work just fine. Here was the original snippet: https://github.com/Qiskit/qiskit/blob/0bc84296f92d8eb9d51f52a4ca117a87804fed31/qiskit/transpiler/target.py#L797-L808

Removed in c7b173f69bd139aaea4002c245b2e6220f189582

eliarbel commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1396,43 +1365,72 @@ impl Target {
+                        let qarg_set: HashSet<PhysicalQubit> = qargs_vec.iter().cloned().collect();
+                        return qargs_vec.iter().all(|qarg| {
+                            qarg.0 <= self.num_qubits.unwrap_or_default()
+                                && qarg_set.len() == qargs_vec.len()
```
I guess you meant for this to be outside the closure.
Also, nitpicking: `self.num_qubits.unwrap_or_default()` should be cached

raynelfss replied:
Cached in c7b173f69bd139aaea4002c245b2e6220f189582

eliarbel reviewed (commented):
Overall looks good, thanks for doing this. I've left a few minor line comments. 

I was also thinking about the following while going over the code:
The flow of the checking logic inside `instruction_supported` could probably be refactored to something along the lines of: 
* If `name` not in in `_gate_name_map` - return false
* If non-global instructions and mismatched `qargs` - return false
* If parameters don't match - return false
* Return true

This seems like a more natural flow that will simplify the code a bit, including I think eliminating one of the qargs checks for the `TargetOperation::Variadic` case.

raynelfss committed:
Fix: Address review comments
- Cache the unwrapping on `num_qubits` in `instruction_supported`.
- Remove duplicate qargs checks.

raynelfss committed:
Fix: Do not check for angles in `GateDirection`.

eliarbel commented on crates/transpiler/src/passes/gate_direction.rs:
```diff
@@ -184,29 +184,13 @@ pub fn fix_direction_coupling_map(
+            PhysicalQubit::new(op_args[1].0),
         ];
 
         // Take this path so Target can check for exact match of the parameterized gate's angle
```
```suggestion
```
We don't need this comment because there is only one path now

raynelfss committed:
Update crates/transpiler/src/passes/gate_direction.rs

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

eliarbel commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1386,60 +1332,98 @@ impl Target {
+    ///
+    /// # Returns
+    ///
+    /// * `true`` if the instruction compatible with the target, `false` if otherwise.
```
```suggestion
    /// * `true` if the instruction is compatible with the target, `false` if otherwise.
```

eliarbel committed:
Apply suggestion from @eliarbel

eliarbel reviewed (dismissed):
This is ready now. Thanks @raynelfss!

raynelfss committed:
Lint: Fix formatting in ConsolidateBlocks
